### PR TITLE
Use crypto files when generating features, errors, query config and visual studio files.

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR is discussed here: https://jira.arm.com/browse/IOTSSL-2801.
Errors, features and the query config use both the crypto and mbedtls files simultaneously, and that's turned on by default. This behaviour can be changed via an additional input parameter.
The visual studio file generating script has an input parameter do enable generating a project with only the crypto .c files - as, by default, the mbedtls source files are used.

### CODE REVIEW

File: scripts/generate_features.pl [hunk 1]
```diff
@@ -36,37 +42,53 @@
 close(FORMAT_FILE);
 
 $/ = $line_separator;
+my %defines_seen;
+my @files = ("$include_dir/config.h");
 
-open(CONFIG_H, "$include_dir/config.h") || die("Failure when opening config.h: $!");
+if( $include_crypto ) {
+    push(@files, "$crypto_include_dir/config.h");
```
Comment (gilles-peskine-arm) [inline scripts/generate_features.pl]: No, for `config.h`, we should not read both files. `config.h` outside the submodule contains the whole configuration.

Applies to `query_config` as well.

File: scripts/generate_visualc_files.pl [hunk 2]
```diff
@@ -18,10 +18,20 @@
...
+    die "Invalid number of arguments" if scalar @ARGV != 1;
+    ($include_crypto) = @ARGV;
+}
+
 my $programs_dir = 'programs';
 my $header_dir = 'include/mbedtls';
 my $source_dir = 'library';
 
+if( $include_crypto ) {
+    $source_dir = 'crypto/library';
```
Comment (gilles-peskine-arm) [inline scripts/generate_visualc_files.pl]: No. If `$include_crypto`, VS must be told to compile files from both `crypto/library` and `library`.

File: scripts/generate_features.pl (hunk 1 above)
Comment (AndrzejKurek) [reply scripts/generate_features.pl]: If so, then there's no change to do in these files... ?

File: scripts/generate_visualc_files.pl (hunk 2 above)
Comment (AndrzejKurek) [reply scripts/generate_visualc_files.pl]: Hmm, I don't think I understand what would that do - why would we want to compile both sources at the same time, in the same project? I thought that the idea is that we can build either the submodule, or mbedTLS sources, both with mbedTLS headers.

Comment (gilles-peskine-arm) [reply scripts/generate_visualc_files.pl]: If we use the submodule, then crypto sources come from `crypto/library` and X509/TLS sources come from `library`.

Comment (AndrzejKurek) [reply scripts/generate_visualc_files.pl]: So should the crypto sources shadow mbedTLS ones, just as in the generate_errors scripts? We would then take all of the sources from crypto and non-duplicated files from mbedtls.

File: scripts/generate_errors.pl [hunk 3]
```diff
@@ -52,14 +54,31 @@
 
 $/ = $line_separator;
 
-my @files = <$include_dir/*.h>;
+my %files;
 
 if( $include_crypto ) {
```
Comment (gilles-peskine-arm) [inline scripts/generate_errors.pl]: This block looks pretty complicated. How about this?
```perl
my @headers = ();
if ($include_crypto) {
    @headers = <$crypto_dir/include/*.h>;
    foreach my $header (<$include_dir/*.h>) {
        my $basename = $header; $basename =~ s!.*/!!;
        push @headers, $header unless -e "$crypto_dir/include/$basename";
    }
} else {
     @headers = <$include_dir/*.h>;
}
```

File: scripts/generate_errors.pl [hunk 4]
```diff
@@ -52,14 +54,31 @@
...
+    $files{$_}++ for (@crypto_headers);
+
+    foreach my $file (@mbedtls_files) {
+        my $stripped_filename = substr($file, rindex($file,"/")+1, length($file)-rindex($file,"/")-1);
+        my $crypto_counterpart = "$crypto_dir/$include_dir/$stripped_filename";
+        if ( exists $files{$crypto_counterpart} ){
+          next;
+        }
+        else{
+          push(@{$files{$file}});
```
Comment (gilles-peskine-arm) [inline scripts/generate_errors.pl]: ITYM `files{$file} = 1`

File: scripts/generate_features.pl (hunk 1 above)
Comment (gilles-peskine-arm) [reply scripts/generate_features.pl]: Indeed.

File: scripts/generate_visualc_files.pl (hunk 2 above)
Comment (gilles-peskine-arm) [reply scripts/generate_visualc_files.pl]: Right.

**Comment by AndrzejKurek** [comment]
Thanks for your input @gilles-peskine-arm. I've updated the PR according to all of your suggestions.

File: scripts/generate_errors.pl [hunk 5]
```diff
@@ -27,6 +30,10 @@
     }
 }
 
+if( $include_crypto ) {
+    -d $crypto_dir or die "Crypto submodule not present\n";
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/generate_errors.pl]: Actually I made a mistake there: `-d` is true on symlinks to directories. So this test is fine.

File: scripts/generate_errors.pl [hunk 6]
```diff
@@ -3,23 +3,28 @@
...
 if( @ARGV ) {
-    die "Invalid number of arguments" if scalar @ARGV != 3;
-    ($include_dir, $data_dir, $error_file) = @ARGV;
+    die "Invalid number of arguments" if scalar @ARGV != 4;
+    ($include_dir, $data_dir, $error_file, $include_crypto) = @ARGV;
 
     -d $include_dir or die "No such directory: $include_dir\n";
     -d $data_dir or die "No such directory: $data_dir\n";
+    if( $include_crypto ) {
+        -d $crypto_dir or die "Crypto submodule not present\n";
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/generate_errors.pl]: Actually, I was wrong: `-d` calls `stat`, not `lstat`, so it accepts symlinks to directories. So `-d $crypto_dir` is fine. Thanks to @k-stachowiak for pointing out that this issue was dangling.

**Pull request merged**